python/dataproc_templates/cassandra/cassandra_to_gcs.py

```diff
@@ -0,0 +1,163 @@
+           dest=constants.CASSANDRA_TO_GCS_OUTPUT_PATH,
+           required=True,
+           help='GCS location for output files'
+       )
+       parser.add_argument(
+           f'--{constants.CASSANDRA_TO_GCS_OUTPUT_SAVEMODE}',
+           dest=constants.CASSANDRA_TO_GCS_OUTPUT_SAVEMODE,
+           required=False,
+           default=constants.OUTPUT_MODE_APPEND,
+           help=(
+               'Output write mode '
+               '(one of: append,overwrite,ignore,errorifexists) '
+               '(Defaults to append)'
+           ),
+           choices=[
+               constants.OUTPUT_MODE_OVERWRITE,
+               constants.OUTPUT_MODE_APPEND,
+               constants.OUTPUT_MODE_IGNORE,
+               constants.OUTPUT_MODE_ERRORIFEXISTS
+           ]
+       )
+       parser.add_argument(
+            f'--{constants.CASSANDRA_TO_GCS_CATALOG}',
+            dest=constants.CASSANDRA_TO_GCS_CATALOG,
+            required=False,
+            default="casscon",
+            help='To provide a name for connection between Cassandra and GCS'
+        )
+       parser.add_argument(
+            f'--{constants.CASSANDRA_TO_GCS_QUERY}',
+            dest=constants.CASSANDRA_TO_GCS_QUERY,
+            required=False,
+            help='Optional query for selective exports'
+        )
+ 
+       known_args: argparse.Namespace
+       known_args, _ = parser.parse_known_args(args)
+ 
+       return vars(known_args)
+ 
+   def run(self, spark: SparkSession, args: Dict[str, Any]) -> None:
+ 
+       logger: Logger = self.get_logger(spark=spark)
+ 
+       # Arguments
+       input_host: str = args[constants.CASSANDRA_TO_GCS_INPUT_HOST]
+       input_keyspace: str = args[constants.CASSANDRA_TO_GCS_INPUT_KEYSPACE]
+       input_table: str = args[constants.CASSANDRA_TO_GCS_INPUT_TABLE]
+       output_format: str = args[constants.CASSANDRA_TO_GCS_OUTPUT_FORMAT]
+       output_mode: str = args[constants.CASSANDRA_TO_GCS_OUTPUT_SAVEMODE]
+       output_location: str = args[constants.CASSANDRA_TO_GCS_OUTPUT_PATH]
+       catalog: str = args[constants.CASSANDRA_TO_GCS_CATALOG]
+       query: str = args[constants.CASSANDRA_TO_GCS_QUERY]
+ 
+       logger.info(
+           "Starting CASSANDRA to GCS spark job with parameters:\n"
+           f"{pprint.pformat(args)}"
+       )
+       
+       # Set configuration to connect to Cassandra by overwriting the spark session
+       spark = (
+        SparkSession
+        .builder
+        .appName("CassandraToGCS")
+        .config(constants.SQL_EXTENSION, constants.CASSANDRA_EXTENSION)
+        .config(f"spark.sql.catalog.{catalog}", constants.CASSANDRA_CATALOG)
+        .config(f"spark.sql.catalog.{catalog}.spark.cassandra.connection.host",input_host)
+        .getOrCreate())
+ 
+       # Read
+       if(not query):
+           input_data = spark.read.table(f"{catalog}.{input_keyspace}.{input_table}")
+       else:
+           input_data = spark.sql(query)  
```

Comment:
As per the current design, either query or {input_keyspace}.{input_table} would be required for execution. That means, if the user provides query, then {input_keyspace}.{input_table} are not to be provided. This is not handled correctly as {input_keyspace}.{input_table} are set as required properties.

Reply:
Modified required properties for keyspace and table to be true only when query is None.
Thanks for pointing that out!

Reply:
Please update README as well.

Reply:
@saumyasinha-google What would happen if all three params: table, keyspace and query is provided by the user?

Reply:
The template will run fine if all three will be provided. However keyspace and table name won't be required properties in such a case but even if they are provided template will run fine.

Reply:
Ideally, it should fail and prompt the user to either provide query or table and keyspace. Please make the necessary changes.

Reply:
Made the changes.